server/grpc_service.go

```diff
@@ -101,6 +116,48 @@ func (s *Server) Tso(stream pdpb.PD_TsoServer) error {
 		if err != nil {
 			return errors.WithStack(err)
 		}
+
+		receiverAddr := getReceiverAddr(stream.Context())
+		if !s.isLocalRequest(receiverAddr) {
+			if redirectStream == nil || lastReceiverAddr != receiverAddr {
+				if cancel != nil {
+					cancel()
+				}
+				client, err := s.getDelegateClient(s.ctx, receiverAddr)
+				if err != nil {
+					return err
+				}
+				redirectStream, cancel, err = s.createTsoRedirectStream(client, receiverAddr)
+				if err != nil {
+					return err
+				}
+				lastReceiverAddr = receiverAddr
+				errCh = make(chan error, 1)
+				go func(redirectStream pdpb.PD_TsoClient, errCh chan error) {
```

Comment:
Would check the goroutine number in your test？ I think it will growth quickly, it only close when stream meet error.